### PR TITLE
Allow graph to be filtered by friend, location and hashtag

### LINKED ISSUE #124: Change `graph` to use `--with`, `--tagged`, and `--in` flags

### PR DESCRIPTION
**Pull request by andypearson**
Resolves #124 

Moved a fair bit of code around here so expecting a few goes on this!

Currently I'm seeing a lot of Rubocop offenses so I'm not sure if we're supposed to be running it at the moment (or I'm running it in the wrong way).

```
08:52 $ bundle exec rubocop
Inspecting 20 files
CCCCCCC.CCCCCCCCCCCC
20 files inspected, 38 offenses detected
```

### CODE REVIEW

File: lib/friends/graph.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+# Graphs activites by month
```
Comment (JacobEvelyn) [inline lib/friends/graph.rb]: Small typo ("activites").

File: lib/friends/introvert.rb [hunk 2]
```diff
@@ -308,32 +295,28 @@ def list_hashtags(from:)
...
+    # @param with [String] the name of a friend to filter by, or nil for
+    #   unfiltered
+    # @param location_name [String] the name of a location to filter by, or nil
+    #   for unfiltered
+    # @param tagged [String] the name of a hashtag to filter by, or nil for
+    #   unfiltered
+    # @return [Hash{String => Integer}]
+    # @raise [FriendsError] if friend, location or hashtag cannot be found or is ambiguous
+    def graph(with:, location_name:, tagged:)
+      return {} if @activities.empty?
```
Comment (JacobEvelyn) [inline lib/friends/introvert.rb]: Why is this necessary?

File: lib/friends/introvert.rb [hunk 3]
```diff
@@ -476,6 +459,38 @@ def elapsed_days
 
     private
 
+    # Filter activites by friend, location and hashtag
```
Comment (JacobEvelyn) [inline lib/friends/introvert.rb]: Same typo (here and in a lot of places). Could you just do a global find-replace?

File: test/introvert_spec.rb [hunk 4]
```diff
@@ -804,82 +804,102 @@ def stub_locations(val)
...
     end
 
-    describe "when friend name is valid" do
-      let(:friend_name) { "George" }
+    describe "Filtering by location" do
+      let(:location_name) { "The Eiffel Tower" }
 
-      it "returns a hash of months and frequencies" do
-        stub_friends(friends) do
+      it "graphs activites with in a location" do
```
Comment (JacobEvelyn) [inline test/introvert_spec.rb]: Remove "with"?

**Comment by JacobEvelyn** [comment]
I just left some comments, but in general this looks really good! I like the refactoring; thanks for doing that.

In terms of Rubocop, it looks like I was using an old version. I just upgraded and found the errors you're seeing so I'll make a a commit to fix them in a minute.

**Comment by JacobEvelyn** [comment]
I just added that commit to update Rubocop and fix all of the linting errors. If you rebase your changes on top of that commit (happy to explain how to do that) the linting issues unrelated to your code should be gone. Let me know if you need any help!

File: lib/friends/introvert.rb (hunk 2 above)
Comment (andypearson) [inline lib/friends/introvert.rb]: Things explode further down when calling `@activities.first` and `@activities.last` so it seemed safer to bail here if there's nothing at all to graph.

**Comment by andypearson** [comment]
I'll do Rubocop and typo fixing when I get home. Would you prefer a single squashed commit to merge?

**Comment by andypearson** [comment]
Travis shows that this branch is good, but I would expect it to tell me that there are Rubocop errors. Is it possible that Travis is running what we think it should?

(Note: I haven't fiddled with overcommit yet)

File: lib/friends/introvert.rb (hunk 2 above)
Comment (JacobEvelyn) [inline lib/friends/introvert.rb]: Ah, gotcha, I missed that. Could you add a comment about this?

**Comment by JacobEvelyn** [comment]
A single squashed commit would be great! Re: Travis, it only runs the minitest specs (`rake test`) and doesn't do any Rubocop analysis. That's an interesting idea though; I've made #131 for that, though as far as I'm aware checking code style in formal tests isn't really a pattern people use.

File: lib/friends/graph.rb (hunk 1 above)
Comment (andypearson) [inline lib/friends/graph.rb]: Fixed!

File: lib/friends/introvert.rb (hunk 2 above)
Comment (andypearson) [inline lib/friends/introvert.rb]: Sure, comment added!

File: test/introvert_spec.rb (hunk 4 above)
Comment (andypearson) [inline test/introvert_spec.rb]: Done!

**Comment by andypearson** [comment]
Single squashed commit should be good to go!

My current team decided that adhering to project coding standards constituted part of the "build", along with passing unit, integration tests and a combined 100% coverage metric. The team contributed to the standards so we were able to settle on a style that everyone was happy with.

For us it made a lot of sense to include Rubocop as part of what makes a new addition to the code base "good". And new additions require a passing build, of course.

I'm happy to actually install overcommit, also :D

**Comment by JacobEvelyn** [comment]
Thanks for the thoughts on "build" rules. Really helpful! I'll be playing around with different ways of enforcing these things (I just noticed [Pronto](https://github.com/mmozuras/pronto), for instance) better in the next few weeks.

**Comment by JacobEvelyn** [comment]
This is out in version 0.23! Thanks again for working on this!
